drivers/serial/uart_hpmicro.c

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright (c) 2022 HPMicro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#define DT_DRV_COMPAT hpmicro_hpm_uart
+
+#include <errno.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/hpmicro_clock_control.h>
+#include <zephyr/irq.h>
+#include <soc.h>
+#include <hpm_common.h>
+#include <hpm_uart_drv.h>
+
+struct uart_hpm_cfg {
+	UART_Type *base;
+	uint32_t parity;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t irq_config_func;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+	const struct pinctrl_dev_config *pincfg;
+	const struct hpm_clock_configure_data uart_clk;
+};
+
+struct uart_hpm_data {
+	uint32_t baud_rate;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t user_cb;
+	void *user_data;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static int uart_hpm_init(const struct device *dev)
+{
+	const struct uart_hpm_cfg *const cfg = dev->config;
+	const struct device *const clk = DEVICE_DT_GET(HPMICRO_CLOCK_CONTROL_NODE);
+	struct uart_hpm_data *const data = dev->data;
+	parity_setting_t parity;
+	hpm_stat_t stat = status_success;
+	uart_config_t config = {0};
+	int ret;
+
+	ret = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (clock_control_configure(clk,
+		(clock_control_subsys_t *) &cfg->uart_clk, NULL) != 0) {
+		return -EIO;
+	}
+
+	uart_default_config((UART_Type *)cfg->base, &config);
+
+	if (clock_control_get_rate(clk,
+		(clock_control_subsys_t *) &cfg->uart_clk, &config.src_freq_in_hz) != 0) {
+		return -EIO;
+	}
+	config.baudrate = data->baud_rate;
+	switch (cfg->parity) {
+	case UART_CFG_PARITY_NONE:
+		parity = parity_none;
+		break;
+	case UART_CFG_PARITY_ODD:
+		parity = parity_odd;
+		break;
+	case UART_CFG_PARITY_EVEN:
+		parity = parity_even;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	config.parity = parity;
+	stat = uart_init((UART_Type *)cfg->base, &config);
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	cfg->irq_config_func(dev);
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+	return 0;
+}
+
+static int uart_hpm_poll_in(const struct device *dev, unsigned char *c)
+{
+	const struct uart_hpm_cfg *const cfg = dev->config;
+
+	if (status_success != uart_receive_byte((UART_Type *)cfg->base, c)) {
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+static void uart_hpm_poll_out(const struct device *dev, unsigned char c)
+{
+	const struct uart_hpm_cfg *const cfg = dev->config;
+
+	uart_send_byte((UART_Type *)cfg->base, c);
+}
+
+static int uart_hpm_err_check(const struct device *dev)
+{
+	int errors = 0;
+
+	return errors;
+}
```

Comment:
Add `ARG_UNUSED(dev)` and change the return statement to just `return 0;`. A comment explaining that HPM UART doesn't support error checking would also be nice